pkg/services/authn/authnimpl/hook/default_org_hook.go

```diff
@@ -0,0 +1,74 @@
+package hook
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/auth/identity"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type DefaultOrgHook struct {
+	cfg         *setting.Cfg
+	logger      log.Logger
+	userService user.Service
+	orgService  org.Service
+}
+
+// ProvideDefaultOrgHook sets the default org for a user after login.
+func ProvideDefaultOrgHook(cfg *setting.Cfg, userService user.Service, orgService org.Service) *DefaultOrgHook {
+	return &DefaultOrgHook{
+		cfg:         cfg,
+		logger:      log.New("authn.default_org_hook"),
+		userService: userService,
+		orgService:  orgService,
+	}
+}
+
+func (h *DefaultOrgHook) SetDefaultOrg(ctx context.Context,
+	currentIdentity *authn.Identity, r *authn.Request, err error) {
+	if err != nil || h.cfg.LoginDefaultOrgId < 1 || currentIdentity == nil {
+		return
+	}
+
+	namespace, identifier := currentIdentity.GetNamespacedID()
+	if namespace != identity.NamespaceUser {
+		return
+	}
+
+	userID, err := identity.IntIdentifier(namespace, identifier)
+	if err != nil {
+		return
+	}
+
+	if !h.validateUsingOrg(ctx, userID, h.cfg.LoginDefaultOrgId) {
+		return
+	}
+
+	cmd := user.SetUsingOrgCommand{UserID: userID, OrgID: h.cfg.LoginDefaultOrgId}
+	if err := h.userService.SetUsingOrg(ctx, &cmd); err != nil {
+		h.logger.Warn("failed to set default org", "err", err)
+	}
```

Comment:
Thinking a bit more about this, does this really have to be its own hook?

We already call `SetUsingOrg` [here](https://github.com/grafana/grafana/blob/a722b2608afcc5768c2afaa6532d52119fce9135/pkg/services/authn/authnimpl/sync/org_sync.go#L119-L127). Could this be part of that?

Reply:
Thanks, I reviewed the `OrgSync` again and the biggest difference is that the `OrgSync` is only executed when the `ClientParams.SyncOrgRoles` is true (for external providers this value is controlled by the `SkipOrgRoleSync` parameter). However, I think setting the default org should be executed regardless of the `SkipOrgRoleSync` setting, so I think we need this separate hook.

https://github.com/grafana/grafana/blob/a722b2608afcc5768c2afaa6532d52119fce9135/pkg/services/authn/authnimpl/sync/org_sync.go#L28-L31